fw/http_stream_sched.c

```diff
@@ -0,0 +1,410 @@
+		 *    according RFC 7540.
+		 */
+		BUG_ON(new_parent == &sched->root);
+		np = container_of(new_parent, TfwStream, sched);
+
+		tfw_h2_stream_sched_remove(np);
+		tfw_h2_stream_sched_remove(stream);
+		tfw_h2_change_stream_weight(stream, new_weight);
+		tfw_h2_add_stream_dep(np, old_parent, false);
+		tfw_h2_add_stream_dep(stream, new_parent, excl);
+	}
+
+}
+
+void
+tfw_h2_sched_stream_enqueue(TfwStream *stream, TfwStreamSchedEntry *parent)
+{
+	/*
+	 * Holds 256 entries of offsets (multiplied by 65536) where nodes
+	 * with weights between 1..256 should go into each entry (except
+	 * for weight=256) is calculated as: round(N / weight), where N
+	 * is adjusted so that the value would become 63*65536 for weight=0.
+	 */
+	static const unsigned offset_tbl[] = {
+		4128768, 2064384, 1376256, 1032192, 825754, 688128, 589824,
+		516096, 458752, 412877, 375343, 344064, 317598, 294912,
+		275251, 258048, 242869, 229376, 217304, 206438, 196608,
+		187671, 179512, 172032, 165151, 158799, 152917, 147456,
+		142371, 137626, 133186, 129024, 125114, 121434, 117965,
+		114688, 111588, 108652, 105866, 103219, 100702, 98304,
+		96018, 93836, 91750, 89756, 87846, 86016, 84261, 82575,
+		80956, 79399, 77901, 76459, 75069, 73728, 72435, 71186,
+		69979, 68813, 67685, 66593, 65536, 64512, 63520, 62557,
+		61623, 60717, 59837, 58982, 58152, 57344, 56558, 55794,
+		55050, 54326, 53620, 52933, 52263, 51610, 50972, 50351,
+		49744, 49152, 48574, 48009, 47457, 46918, 46391, 45875,
+		45371, 44878, 44395, 43923, 43461, 43008, 42565, 42130,
+		41705, 41288, 40879, 40478, 40085, 39700, 39322, 38951,
+		38587, 38229, 37879, 37534, 37196, 36864, 36538, 36217,
+		35902, 35593, 35289, 34990, 34696, 34406, 34122, 33842,
+		33567, 33297, 33030, 32768, 32510, 32256, 32006, 31760,
+		31517, 31279, 31043, 30812, 30583, 30359, 30137, 29919,
+		29703, 29491, 29282, 29076, 28873, 28672, 28474, 28279,
+		28087, 27897, 27710, 27525, 27343, 27163, 26985, 26810,
+		26637, 26466, 26298, 26131, 25967, 25805, 25645, 25486,
+		25330, 25175, 25023, 24872, 24723, 24576, 24431, 24287,
+		24145, 24004, 23866, 23729, 23593, 23459, 23326, 23195,
+		23066, 22938, 22811, 22686, 22562, 22439, 22318, 22198,
+		22079, 21962, 21845, 21730, 21617, 21504, 21393, 21282,
+		21173, 21065, 20958, 20852, 20748, 20644, 20541, 20439,
+		20339, 20239, 20140, 20043, 19946, 19850, 19755, 19661,
+		19568, 19475, 19384, 19293, 19204, 19115, 19027, 18939,
+		18853, 18767, 18682, 18598, 18515, 18432, 18350, 18269,
+		18188, 18109, 18030, 17951, 17873, 17796, 17720, 17644,
+		17569, 17495, 17421, 17348, 17275, 17203, 17132, 17061,
+		16991, 16921, 16852, 16784, 16716, 16648, 16581, 16515,
+		16449, 16384, 16319, 16255, 16191, 16128
+	};
+	TfwStreamSchedEntryAnchor *anchor;
+
+	stream->link.offset = offset_tbl[stream->weight - 1] + stream->link.deficit;
+	stream->link.deficit = stream->link.offset % 4128768;
+	stream->link.offset = ((stream->link.offset / 65536) % TFW_STREAM_SCHED_ENTRY_COUNT);
```

Comment:
This looks like a dummy hash function just getting 6 bits in the middle of the `deficit` value ignoring the most significant (in both the meaning!) bits of `deficit`. It seems this is a math error: in the original formula (https://www.mew.org/~kazu/material/2015-http2-priority2.pdf slide 10):
```
deficit = (last_deficit + constant / weight) % constant2
offset = (last_deficit + constant / weight) / constant2
```
the first calculation is just a wrapping using exactly less significant part of `deficit`, while the second uses exactly the most significant half of the value for the offset. The formula doesn't ignore any single bit, but the code does.

This calculation is important because the most close deficit values (by the most significant bits) appear in the same anchor buckets. In the code a deficit with zero most significant bit may be placed in the same bucket with deficit with the bit set.